### PR TITLE
Add support for overriding the archive filename

### PR DESCRIPTION
**Pull request by JoakimLofgren**
Instead of running a shell command to rename the archive I want to be able to specify a option to do it straight away.

`composer archive --format=zip --filename=my-package` would create a archive `my-package.zip`

### CODE REVIEW

File: tests/Composer/Test/Package/Archiver/ArchiveManagerTest.php [hunk 1]
```diff
@@ -55,6 +55,23 @@ public function testArchiveTar()
...
+
+        $package = $this->setupPackage();
+
+        $this->manager->archive($package, 'tar', $this->targetDir, 'override');
+
+        $target = $this->targetDir.'/override.tar';
+        $this->assertFileExists($target);
+
+        $tmppath = sys_get_temp_dir().'/composer_archiver/'.$this->manager->getPackageFilename($package);
+        $this->assertFileNotExists($tmppath);
```
Comment (JoakimLofgren) [inline tests/Composer/Test/Package/Archiver/ArchiveManagerTest.php]: I copied this from the `testArchiveTar` test.
Looking at the [ArchiveManager](https://github.com/composer/composer/blob/master/src/Composer/Package/Archiver/ArchiveManager.php#L157) wouldn't this always assert to true?
Considering that the $tempTarget is using `uniqid()` and the directories doesn't seem to match anymore either ?

Commenting out the clean up steps in the ArchiveManager seems to still give the same test results.

File: src/Composer/Package/Archiver/ArchiveManager.php [hunk 2]
```diff
@@ -123,6 +124,9 @@ public function archive(PackageInterface $package, $format, $targetDir)
 
         $filesystem = new Filesystem();
         $packageName = $this->getPackageFilename($package);
+        if ($filename) {
+            $packageName = $filename;
```
Comment (stof) [inline src/Composer/Package/Archiver/ArchiveManager.php]: you could avoid building the default filename entirely in thi case

File: tests/Composer/Test/Package/Archiver/ArchiveManagerTest.php (hunk 1 above)
Comment (Seldaek) [inline tests/Composer/Test/Package/Archiver/ArchiveManagerTest.php]: Yes that seems like a useless assertion indeed. It's kinda hard to test for given the uniqid so I'd just remove the assertion altogether.

**Comment by Seldaek** [comment]
Ping @JoakimLofgren - any chance you wrap up those few things that were commented upon above?

**Comment by JoakimLofgren** [comment]
Oh sorry, had kinda forgotten about this :)
I'll take a look at it tomorrow.

**Comment by JoakimLofgren** [comment]
Wasn't this implemented by #4479 ?
So this PR can be closed ?

**Comment by Seldaek** [comment]
Oh god yes sorry I missed that :)

**Comment by JoakimLofgren** [comment]
No problem, at least someone fixed the filename feature ;)

**Pull request closed without merging**